### PR TITLE
MAILBOX-398 Backend is responsible of RECENT search

### PR DESCRIPTION
**Pull request by chibenwa**
A search query should not require stored information to be built.

This refactoring helps simplifying search related code

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[6fd9b1fb4720d191a1f305d5c02acaa0010c401b] [ERROR]   MemoryTerminationSubscriberTest.dynamicListeningEventsShouldGetOnlyNewEvents 
[6fd9b1fb4720d191a1f305d5c02acaa0010c401b] Expecting:
[6fd9b1fb4720d191a1f305d5c02acaa0010c401b]   <[Cancelled(TaskAggregateId(TaskId{value=f5ead265-fd23-4af3-aa79-0fb2ca4a2783}),EventId(42),None)]>
[6fd9b1fb4720d191a1f305d5c02acaa0010c401b] to contain exactly (and in same order):
[6fd9b1fb4720d191a1f305d5c02acaa0010c401b]   <[Failed(TaskAggregateId(TaskId{value=ab92a44e-7e86-4ffe-9e5c-6a7c3dc13702}),EventId(42),None,None,None),
[6fd9b1fb4720d191a1f305d5c02acaa0010c401b]     Cancelled(TaskAggregateId(TaskId{value=f5ead265-fd23-4af3-aa79-0fb2ca4a2783}),EventId(42),None)]>
[6fd9b1fb4720d191a1f305d5c02acaa0010c401b] but could not find the following elements:
[6fd9b1fb4720d191a1f305d5c02acaa0010c401b]   <[Failed(TaskAggregateId(TaskId{value=ab92a44e-7e86-4ffe-9e5c-6a7c3dc13702}),EventId(42),None,None,None)]>
```

test this please

**Comment by chibenwa** [comment]
```

[844680bef87b66083d94de4fdd73c7b3116de335] [ERROR] src/main/java/org/apache/james/mailbox/model/SearchQuery.java:[26,8] (imports) UnusedImports: Unused import - java.util.HashSet.
[844680bef87b66083d94de4fdd73c7b3116de335] [ERROR] src/main/java/org/apache/james/mailbox/model/SearchQuery.java:[28,8] (imports) UnusedImports: Unused import - java.util.Set.
```

**Comment by chibenwa** [comment]
```
[813dc07c407e1abeb4c871030527195fc83a8775] [ERROR] Failures: 
[813dc07c407e1abeb4c871030527195fc83a8775] [ERROR]   SearchUtilsTest.testShouldMatchSeenRecentSet:508 
[813dc07c407e1abeb4c871030527195fc83a8775] Expecting:
[813dc07c407e1abeb4c871030527195fc83a8775]  <false>
[813dc07c407e1abeb4c871030527195fc83a8775] to be equal to:
[813dc07c407e1abeb4c871030527195fc83a8775]  <true>
[813dc07c407e1abeb4c871030527195fc83a8775] but was not.
[813dc07c407e1abeb4c871030527195fc83a8775] [ERROR]   SearchUtilsTest.testShouldMatchSeenRecentUnSet:586 
[813dc07c407e1abeb4c871030527195fc83a8775] Expecting:
[813dc07c407e1abeb4c871030527195fc83a8775]  <false>
[813dc07c407e1abeb4c871030527195fc83a8775] to be equal to:
[813dc07c407e1abeb4c871030527195fc83a8775]  <true>
[813dc07c407e1abeb4c871030527195fc83a8775] but was not.
```

**Comment by chibenwa** [comment]
`RabbitMQEventBusTest$LifeCycleTest$SingleEventBus.dispatchShouldWorkAfterRestartForNewKeyRegistration » RetryExhausted`

Unrelated

test this please

**Comment by chibenwa** [comment]
```
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Location: /org/apache/james/imap/scripts/SearchAtoms.test:1703
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] LastClientMsg: A79 SEARCH NEW
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Expected: '\* SEARCH 3 4 8 9 13 14 18 19'
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Actual   : '* SEARCH'
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] 
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] [ERROR] testSearchAtomsITALY  Time elapsed: 1.616 s  <<< ERROR!
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] org.apache.james.mpt.protocol.ProtocolSession$InvalidServerResponseException: 
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] 
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Location: /org/apache/james/imap/scripts/SearchAtoms.test:1703
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] LastClientMsg: A79 SEARCH NEW
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Expected: '\* SEARCH 3 4 8 9 13 14 18 19'
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Actual   : '* SEARCH'
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] 
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] [ERROR] testSearchAtomsKOREA  Time elapsed: 1.57 s  <<< ERROR!
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] org.apache.james.mpt.protocol.ProtocolSession$InvalidServerResponseException: 
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] 
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Location: /org/apache/james/imap/scripts/SearchAtoms.test:1703
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] LastClientMsg: A79 SEARCH NEW
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Expected: '\* SEARCH 3 4 8 9 13 14 18 19'
[ca926b68a5d4ef14c47556157c1a3e7fcf033c0f] Actual   : '* SEARCH'
```

Will need to investigate

**Comment by chibenwa** [comment]
I found out `SelectedMailboxImpl` recent handling is mosty uncorrelated with the backend storage, and maintains RECENT flags on a per session basis, what regarding what's presented in this ticket, is not what we might expect from the RFC.

We might very well rework RECENT handling to do DB reads.

Thoughts @rouazana ?

File: mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/SearchAtoms.test [hunk 1]
```diff
@@ -1716,15 +1713,9 @@ S: A82 OK SEARCH completed\.
 C: A83 SEARCH 17:*
 S: \* SEARCH 17 18 19
 S: A83 OK SEARCH completed\.
-C: A84 SEARCH OLD
-S: \* SEARCH
-S: A84 OK SEARCH completed\.
 C: A85 SEARCH OR ANSWERED FLAGGED
 S: \* SEARCH 1 2 3 4 5 6 7 8 9 10 11 12 13 14
 S: A85 OK SEARCH completed\.
-C: A86 SEARCH RECENT
```
Comment (rouazana) [inline mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/SearchAtoms.test:1725]: isn't it worth to replace these 3 IMAP calls by there actual results?

File: protocols/imap/src/main/java/org/apache/james/imap/processor/SearchProcessor.java [hunk 2]
```diff
@@ -265,10 +265,6 @@ private ModSeq findHighestModSeq(MailboxSession session, MessageManager mailbox,
 
     private SearchQuery toQuery(SearchKey key, ImapSession session) throws MessageRangeException {
         final SearchQuery result = new SearchQuery();
-        final SelectedMailbox selected = session.getSelected();
-        if (selected != null) {
-            result.addRecentMessageUids(selected.getRecent());
```
Comment (rouazana) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/SearchProcessor.java:270]: there is no more way to search for recent emails, or did I miss something?

Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/SearchProcessor.java:270]: There is: by querying directly what's stored.

File: mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/SearchAtoms.test (hunk 1 above)
Comment (chibenwa) [reply mpt/impl/imap-mailbox/core/src/main/resources/org/apache/james/imap/scripts/SearchAtoms.test:1725]: They return nothing, that's a commit to contnue the build process, but I already know if we want to move further on this topic that we might need a rework of RECENT handling in the IMAP layer.

**Comment by chibenwa** [comment]
(blind as we can't review build error) test this please

**Comment by chibenwa** [comment]
`UidSearch.testSearchAtomsITALY:59 » InvalidServerResponse`

**Comment by chibenwa** [comment]
Appart from the RFC conformity, there is a *bug* behind that topic: On top of ElasticSearch, RECENT Search is de-correlated  from the IMAP client view.

IMAP RECENT search will not return some results that the client IMAPSession presents as RECENT

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Aborted by JamesAdmin

Rude.

test this please

**Comment by Arsnael** [comment]
> Aborted by JamesAdmin
> 
> Rude.

Not rude, likely the build that was running and stuck for more than a week on the CI, so yeah... forced abortion :trollface:

**Comment by Arsnael** [comment]
Log ouput lost

test this please

**Comment by chibenwa** [comment]
test this please (build rescheduled)

**Comment by chibenwa** [comment]
404 test this please

**Comment by chibenwa** [comment]
```
[72da882ce0a05c678d56abbb2c521380c1d1c52b] HEAD is now at 72da882... fixup! IMAP APPEND messages are not considered as RECENT
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Switched to a new branch 'SHA1_BRANCH'
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Switched to branch 'master'
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Your branch is up-to-date with 'apache/master'.
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Switched to a new branch 'MERGE-recent-search-b8b2ef85-9ad1-4060-8060-39dd9a9248dd'
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Auto-merging protocols/imap/src/main/java/org/apache/james/imap/processor/SearchProcessor.java
[72da882ce0a05c678d56abbb2c521380c1d1c52b] CONFLICT (content): Merge conflict in protocols/imap/src/main/java/org/apache/james/imap/processor/SearchProcessor.java
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Auto-merging mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java
[72da882ce0a05c678d56abbb2c521380c1d1c52b] CONFLICT (content): Merge conflict in mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Auto-merging mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Auto-merging mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java
[72da882ce0a05c678d56abbb2c521380c1d1c52b] CONFLICT (content): Merge conflict in mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java
[72da882ce0a05c678d56abbb2c521380c1d1c52b] Automatic merge failed; fix conflicts and then commit the result.
```

rebase needed

**Comment by chibenwa** [comment]
Got a :green_apple: build, we now know what it takes to fix this.

I'm closing the issue as I don't plan to dig further into it

**Pull request closed without merging**